PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/cloudshell/actions/StartAzureCloudShellAction.kt

```diff
@@ -0,0 +1,194 @@
+                                return "${value.subscriptionName} (${value.subscriptionId})"
+                            }
+
+                            return super.getTextFor(value)
+                        }
+
+                        override fun onChosen(selectedValue: SubscriptionDetail, finalChoice: Boolean): PopupStep<*>? {
+                            doFinalStep {
+                                startAzureCloudShell(project, azureManager, selectedValue)
+                            }
+                            return PopupStep.FINAL_CHOICE
+                        }
+                    }
+
+                    val popup = JBPopupFactory.getInstance().createListPopup(step)
+                    popup.showCenteredInCurrentWindow(project)
+                }
+            }
+        }.queue()
+    }
+
+    private fun startAzureCloudShell(project: Project, azureManager: AzureManager, subscriptionDetail: SubscriptionDetail) {
+        try {
+            val authManager = AdAuthManager.getInstance()
+            val tokenCredentials = RefreshableTokenCredentials(authManager, subscriptionDetail.tenantId)
+
+            provisionAzureCloudShell(project, azureManager, tokenCredentials, subscriptionDetail)
+        } catch (e: AuthenticationException) {
+            // Failed to authenticate....
+            AzureCloudShellNotifications.notify(project,
+                    "Azure",
+                    "Failed to authenticate Azure Cloud Shell",
+                    "Authentication was unsuccessful.",
+                    NotificationType.WARNING,
+                    null)
+        }
+    }
+
+    private fun provisionAzureCloudShell(project: Project, azureManager: AzureManager, tokenCredentials: ServiceClientCredentials, subscriptionDetail: SubscriptionDetail) {
+        ProgressManager.getInstance().run(object : Task.Backgroundable(project, this.templatePresentation.text) {
+            override fun run(indicator: ProgressIndicator) {
+                indicator.fraction = 0.05
+                indicator.text = "Retrieving cloud shell preferences..."
+
+                // Use custom API's
+                val retrofitClient = azureManager.getRetrofitClient(
+                        azureManager.environment.azureEnvironment,
+                        AzureEnvironment.Endpoint.RESOURCE_MANAGER,
+                        CloudConsoleService::class.java,
+                        tokenCredentials)
+
+                val consoleUserSettingsResponse = retrofitClient.userSettings().execute()
+                if (consoleUserSettingsResponse.isSuccessful) {
+                    indicator.fraction = 0.35
+                    indicator.text = "Provisioning cloud shell..."
+
+                    // We can provision a shell!
+                    val provisionParameters = CloudConsoleProvisionParameters(
+                            CloudConsoleProvisionParameters.Properties(consoleUserSettingsResponse.body()!!.properties!!.preferredOsType!!))
```

Comment:
Should not we check values for all those force unwrapping? Here and below.

Reply:
Good idea. Updated.